tests/suites/test_suite_pkcs11_client.function

```diff
@@ -0,0 +1,483 @@
+        {CKA_VERIFY, &ck_true, sizeof( ck_true )},
+    };
+    CK_ATTRIBUTE private_attributes[] = {
+        {CKA_DECRYPT, &ck_true, sizeof( ck_true )},
+        {CKA_SIGN, &ck_true, sizeof( ck_true )},
+    };
+    CK_ULONG ck_rsa_key_size = RSA_KEY_SIZE_BITS;
+    unsigned char ecParams[16];
+    size_t ecParams_length;
+
+    switch( key_type )
+    {
+#if defined(MBEDTLS_RSA_C)
+    case MBEDTLS_PK_RSA:
+        mechanism.mechanism = CKM_RSA_PKCS_KEY_PAIR_GEN;
+        public_attributes[0].type = CKA_MODULUS_BITS;
+        public_attributes[0].pValue = &ck_rsa_key_size;
+        public_attributes[0].ulValueLen = sizeof( ck_rsa_key_size );
+        break;
+#endif /* MBEDTLS_RSA_C */
+    default:
+        test_fail( "Unsupported key type in test data", __LINE__, __FILE__ );
+        break;
+    }
+
+    return( C_GenerateKeyPair( hSession,
+                               &mechanism,
+                               public_attributes,
+                               ARRAY_LENGTH( public_attributes ),
+                               private_attributes,
+                               ARRAY_LENGTH( private_attributes ),
+                               phPublicKey, phPrivateKey ) );
+exit:
+    /* Shouldn't happen except if there's a test error (e.g. trying to
+       use a curve that isn't compiled in). */
+    return( -1 );
+}
+
+
+#endif /* MBEDTLS_PK_C */
+
+/* END_HEADER */
+
+/* BEGIN_DEPENDENCIES
+ * depends_on:MBEDTLS_PKCS11_CLIENT_C
+ * END_DEPENDENCIES
+ */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PK_C:MBEDTLS_SHA256_C */
+void pk_generate_sign( int key_type )
+{
+    mbedtls_pk_context pkcs11_ctx;
+    mbedtls_pk_context transparent_ctx;
+    CK_SESSION_HANDLE hSession = CK_INVALID_HANDLE;
+    CK_OBJECT_HANDLE hPublicKey = CK_INVALID_HANDLE;
+    CK_OBJECT_HANDLE hPrivateKey = CK_INVALID_HANDLE;
+    unsigned char hash_value[32] = "Fake hash, it doesn't matter....";
+    unsigned char sig_buffer[RSA_KEY_SIZE_BYTES];
+    size_t sig_length = sizeof( sig_buffer );
+
+    mbedtls_pk_init( &pkcs11_ctx );
+    mbedtls_pk_init( &transparent_ctx );
+
+    /* Initialize cryptoki and generate a key in the token */
+    hSession = pkcs11_init( );
+    TEST_ASSERT( hSession != CK_INVALID_HANDLE );
+
+    CK_ASSERT( pkcs11_generate_key( key_type,
+                                    hSession,
+                                    &hPublicKey, &hPrivateKey ) );
+    TEST_ASSERT( hPublicKey != CK_INVALID_HANDLE );
+    TEST_ASSERT( hPrivateKey != CK_INVALID_HANDLE );
+
+    /* Prepare the mbed TLS contexts */
+    TEST_ASSERT( mbedtls_pk_setup( &transparent_ctx,
+                                   mbedtls_pk_info_from_type( key_type ) ) == 0 );
+    TEST_ASSERT( mbedtls_pk_setup_pkcs11( &pkcs11_ctx,
+                                          hSession,
+                                          hPublicKey,
+                                          hPrivateKey ) == 0 );
+
+    /* Retrieve the public key from the token */
```

Comment:
Can we extract this as a separate function? We might even consider making it (or at least the part concerning public keys) part of `pkcs11_client.c`: Currently,  `pkcs11_client.c` only has functionality for importing a transparent key into a PCKS11 token, but none going in the other direction - which at least for public keys would be very helpful.

Reply:
I consider this outside the scope of the project. I'm not against having it, but at the moment the focus of the PKCS#11 is to support the use of a private key in a PKCS#11 module in TLS.

Do you think this should be a PKCS#11-specific function, or a generic function that all opaque key engines may implement? For a generic function, I'd rather wait until a future version when we have a better idea of what formats are useful and what metadata extraction is useful. For PKCS#11, I'd rather wait until there's a generic API.

Reply:
I meant a function building solely on PKCS#11, similar to the rest of the `pkcs11_client` module (actually, I also didn't yet understand what the generic alternative you mentioned would be). 

From my understanding, the `pkcs11_client` module is meant to provide the glue between PKCS#11 tokens and our PK interface, currently concretely consisting of: 
1. Functionality to move a transparent PK context to the PKCS#11 token.
2. A single opaque PK context implementation that can be created from a PKCS#11 token.

What I'd like to add is

3. Functionality to set up a transparent public PK context from a PKCS#11 token.

and potentially, if only for testing purposes, 

4. Functionality to set up a transparent private PK context from a PKCS#11 token, if possible.